pkg/cmd/infra/router/template.go

```diff
@@ -565,3 +566,78 @@ func (o *TemplateRouterOptions) blueprintRoutes(routeclient *routeclientset.Clie
 
 	return blueprints, nil
 }
+
+// makeTLSConfig checks whether metrics TLS is configured and
+// if so returns a tls.Config whose certificate is automatically
+// reloaded on the given period.
+func makeTLSConfig(reloadPeriod time.Duration) (*tls.Config, error) {
+	certFile := env("ROUTER_METRICS_TLS_CERT_FILE", "")
+	if len(certFile) == 0 {
+		return nil, nil
+	}
+	keyFile := env("ROUTER_METRICS_TLS_KEY_FILE", "")
+
+	// Load the initial certificate contents.
+	certBytes, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return nil, err
+	}
+	keyBytes, err := ioutil.ReadFile(keyFile)
+	if err != nil {
+		return nil, err
+	}
+	certificate, err := tls.X509KeyPair(certBytes, keyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Safely reload the certificate on a fixed period for simplicity.
+	ticker := time.NewTicker(reloadPeriod)
+	var lock sync.Mutex
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				latestCertBytes, err := ioutil.ReadFile(certFile)
+				if err != nil {
+					log.Error(err, "failed to read certificate file", "cert", certFile, "key", keyFile)
+					break
+				}
+				latestKeyBytes, err := ioutil.ReadFile(keyFile)
+				if err != nil {
+					log.Error(err, "failed to read key file", "cert", certFile, "key", keyFile)
+					break
+				}
+
+				if bytes.Equal(latestCertBytes, certBytes) && bytes.Equal(latestKeyBytes, keyBytes) {
+					// Nothing changed, try later.
+					break
+				}
+
+				// Something changed, reload the certificate.
+				certBytes = latestCertBytes
+				keyBytes = latestKeyBytes
+				latest, err := tls.X509KeyPair(certBytes, keyBytes)
+				if err != nil {
```

Comment:
This means the router will not keep trying to reload the certificate if `tls.X509KeyPair` fails.  Now that I think about it, that is probably a good thing—`tls.X509KeyPair` is unlikely to have a transient error, so if loading fails, it probably isn't worth retrying unless the certificate changes.

Reply:
Good catch, this was _not_ my intent.

In the absence of analyzing the error to establish whether it's actually transient, I would prefer to be defensive and retry. Also, I find it potentially confusing to quietly cache the state as if we had succeeded.

Objection to me predicating the certBytes and keyBytes assignment on successful loading?

Reply:
No objection.  It's a tradeoff between some log noise from retrying when the certificate is invalid, or forcing the admin to delete the secret so a new one gets regenerated in case of a transient failure.  I don't feel strongly one way or the other.